cassiopeia/solver/solver_utilities.py

```diff
@@ -0,0 +1,172 @@
+    
+    Args:
+        vec1: The first character vector
+        vec2: The second character vector
+        missing_char: The character representing missing values
+
+    Returns:
+        A list representing the character vector of the lca
+
+    """
+    assert len(vec1) == len(vec2)
+    lca_vec = [0] * len(vec1)
+    for i in range(len(vec1)):
+        if vec1[i] == missing_char and vec2[i] != missing_char:
+            lca_vec[i] = vec2[i]
+        if vec2[i] == missing_char and vec1[i] != missing_char:
+            lca_vec[i] = vec1[i]
+        if vec1[i] == vec2[i] and vec1[i] != "0":
+            lca_vec[i] = vec1[i]
+    return lca_vec
+
+
+def infer_ancestral_characters(T: nx.DiGraph, node: int, char_map: Dict[int, List[str]], missing_char: str):
+    """Annotates the character vectors of the internal nodes of a reconstructed
+    network from the samples, obeying Camin-Sokal Parsimony.
+
+    For an internal node, annotates that node's character vector to be the LCA
+    of its daughters character vectors. Annotates from the samples.
+
+    Args:
+        T: A networkx DiGraph object representing the tree
+        node: The node whose state is to be inferred
+        char_map: A dictionary that maps nodes to their character vectors
+        missing_char: The character representing missing values
+
+    Returns:
+        None, annotates char_map dictionary with node/character vector pairs
+
+
+    """
+    if T.out_degree(node) == 0:
+        return
+    vecs = []
+    for i in T.successors(node):
+        infer_ancestral_characters(T, i, char_map, missing_char)
+        vecs.append(char_map[i])
+    assert len(vecs) == 2
+    lca_characters = get_lca_characters(vecs[0], vecs[1], missing_char)
+    char_map[node] = lca_characters
+    return
+
+
+def collapse_edges(T: nx.DiGraph, node: int, char_map: Dict[int, List[str]]):
+    """A helper function to help collapse mutationless edges in a tree.
+
+    Collapses an edge if the character vector of the parent node is identical
+    to its daughter, removing the identical daughter and creating edges between
+    the parent and the daughter's children. Does not collapse at the level of 
+    the samples. Can create multifurcating trees from strictly binary trees.
+
+    Args:
+        T: A networkx DiGraph object representing the tree
+        node: The node whose state is to be inferred
+        char_map: A dictionary that maps nodes to their character vectors
+
+    Returns:
+        None, operates on the tree destructively
+    """
+    if T.out_degree(node) == 0:
+        return
+    to_remove = []
+    to_collapse = []
+    for i in T.successors(node):
+        to_collapse.append(i)
+    for i in to_collapse:
+        collapse_edges(T, i, char_map)
+        if char_map[i] == char_map[node]:
+            for j in T.successors(i):
+                T.add_edge(node, j)
+            to_remove.append(i)
+    for i in to_remove:
+        T.remove_node(i)
+    return
```

Comment:
Is this supposed to return a collapsed tree, or operate in place? Let's make this more clear in the documentation. And if it does operate on the tree in place, let's remove the return statement.